_posts/2021-11-18-keras-updates/keras-updates.Rmd

```diff
@@ -0,0 +1,159 @@
+
+*Now, what has been going on with Keras?*
+
+## State of the ecosystem
+
+Let us start with a characterization of the ecosystem, and a few words on history.
+
+In this post, when we say *Keras*, we mean R -- as opposed to Python -- *Keras*[^1]. Now, this immediately translates to the R package `keras`. But `keras` alone wouldn't get you far. While `keras` provides the high-level functionality -- neural network layers, optimizers, workflow management, and more -- the basic data structure operated upon, *tensors*, lives in `tensorflow`. Thirdly, as soon as you'll need to perform less-then-trivial pre-processing, or can no longer keep the whole training set in memory because of its size, you'll want to look into `tfdatasets`.
+
+[^1]: Originally, both Keras and Tensorflow are Python libraries, bound to by R at runtime. In practice, this means that those Python libraries have to be present on the R user's machine. This is in contrast to R `torch` , which has no Python dependencies. (Instead, it directly binds to a C++ library.)
+
+So it is these three packages -- `tensorflow`, `tfdatasets`, and `keras` -- that should be understood by "Keras" in the current context[^2]. (The R-Keras ecosystem, on the other hand, is quite a bit bigger. But other packages, such as `tfruns` or `cloudml`, are more decoupled from the core.)
+
+[^2]: Definitely, `tfautograph` belongs in this group as well, providing even-lower-level support for already-low-level `tensorflow`. But normally, the user will not have to call its functions directly.
+
+Matching their tight integration, the aforementioned packages tend to follow a common release cycle, itself dependent on the underlying Python library, [TensorFlow](http://tensorflow.org). For each of `tensorflow`, `tfdatasets`, and `keras` , the current CRAN version is 2.7.0, reflecting the corresponding Python version. The synchrony of versioning between the two Kerases, R and Python, seems to indicate that their fates had developed in similar ways. Nothing could be less true, and knowing this can be helpful.
+
+In R, between present-from-the-outset packages `tensorflow` and `keras`, responsibilities have always been distributed the way they are now: `tensorflow` providing indispensable basics, but often, remaining completely transparent to the user; `keras` being the thing you use in your code. In fact, it is possible to train a Keras model without ever consciously using `tensorflow`.
+
+On the Python side, things have been undergoing significant changes, ones where, in some sense, the latter development has been inverting the first. In the beginning, TensorFlow and [Keras](keras.io) were separate libraries, with TensorFlow providing a backend -- one among several -- for Keras to make use of. At some point, Keras code got incorporated into the TensorFlow codebase. Finally (as of today), following an extended period of slight confusion, Keras got moved out again, and has started to -- again -- considerably grow in features.
```

Comment:
Since you're alluding to what were (presumably) big changes in the Python version, it might be helpful to provide links to a _something_ where this happened (or a good post on it), for those who want more detail.

Reply:
The changes are actually what's following - not alluding to technical changes here, but procedural ones